cpmpy/expressions/globalconstraints.py

```diff
@@ -193,7 +193,26 @@ def decompose(self):
         return [(var1 == var2).implies(var1 == 0) for var1, var2 in all_pairs(self.args)], []
 
     def value(self):
-        vals = [a.value() for a in self.args if a.value() != 0]
+        vals = [argval(a) for a in self.args if argval(a) != 0]
+        return len(set(vals)) == len(vals)
+
+
+class AllDifferentExceptN(GlobalConstraint):
+    """
+    All arguments except those equal to n have a distinct value.
+    """
+    def __init__(self, arr, n):
+        flatarr = flatlist(arr)
+        super().__init__("alldifferent_except_n", [flatarr, n])
+
+    def decompose(self):
+        if self.args[1] == 0:
+            return [AllDifferentExcept0(self.args[0])], []
+        # equivalent to (var1 == n) | (var2 == n) | (var1 != var2)
+        return [(var1 == var2).implies(var1 == self.args[1]) for var1, var2 in all_pairs(self.args[0])], []
+
+    def value(self):
+        vals = [argval(a) for a in self.args[0] if argval(a) != argval(self.args[1])]
```

Comment:
why different .value() than the alldifferent_except0?

Which of the 2 is more efficient? which should keep in both the more efficient one!

Reply:
I don't see what you mean, it's the same?

Reply:
Ah idd, probably looked at the decomposed() before